client/extensions/hello-dolly/state/reducer.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import { HELLO_DOLLY_NEXT_LYRIC } from './action-types';
+import {
+	ROUTE_SET,
+	SECTION_SET,
+	SITE_SETTINGS_SAVE,
+} from 'state/action-types.js';
+
+export default function lyricIndex( state = 0, action ) {
+	switch ( action.type ) {
+		case ROUTE_SET:
+		case SECTION_SET:
+		case SITE_SETTINGS_SAVE:
+		case HELLO_DOLLY_NEXT_LYRIC:
+			return advanceToNextLyric( state );
+		default:
+			return state;
+	}
+}
+
+function advanceToNextLyric( index ) {
+	return index + 1;
```

Comment:
Hmm, if we're really worried about this, I guess we'd need the length of the lyrics we're using, but that's not really accessible here. We could put the length in as state, but if we do, I wonder if we might as well just put all the lyrics in as state then.